### DISCUSSION TITLE
Large Tiff File Handling - High Memory usage - Combining a Tiff file.

### DISCUSSION THREAD

**Discussion by M4nju** [dlemstra/Magick.NET#736] [Help]
Hello,

i am struggeling with handling large Tiff files. I have two Problems of which one of them i already have an solution. 

1. Splitting a multipage tiff into single page tiff files.
2. Combining single page tiff files into one multipage tiff.

The big problem is, if i use the ImageMagickCollection to solve my issue is, that the memory usage is enormos. (The whole file will be red into the RAM and because the memory is not enough, the hdd/sdd will be used as memory).

My Solution for splitting the multipage Tiff is simply only opening one frame at a time and saving this frame:

            MagickReadSettings readSettings = new MagickReadSettings();
            for (int i = 0; i < frameCount; i++)
            {
                readSettings.FrameIndex = i;
                using (MagickImageCollection multipageTif = new MagickImageCollection())
                {
                    multipageTif.Read(@"E:\Testfile.tif", readSettings);
                    if (multipageTif[0] is MagickImage currentPage)
                    {
                        currentPage.Write(Path.Combine(@"E:\Result\", Guid.NewGuid() + ".tif"));
                    }
                }
            }

Like this only one page of the multipage tiff is inside the RAM at a time. Is there something like this for combining a multipage tiff? So i would take a tiff and "SaveAdd" it to an existing multipage tiff. So it would just be appended to the existing tiff.

Thanks for any kind of help.

Kind Regards
Manju

**Comment by M4nju** [comment]
This is how i would do it with native c# code. In this example i open a multipage tiff and save it frame by frame to a file. With this solution not all data will be loaded into the ram, but i am missing the advantages of ImageMagick:

```    ImageCodecInfo info = null;
            foreach (ImageCodecInfo ice in ImageCodecInfo.GetImageEncoders())
                if (ice.MimeType == "image/tiff")
                    info = ice;

            System.Drawing.Imaging.Encoder enc = System.Drawing.Imaging.Encoder.SaveFlag;
            EncoderParameters encoderParameters = new EncoderParameters(1);
            encoderParameters.Param[0] = new EncoderParameter(enc, (long)EncoderValue.MultiFrame);

            using (Bitmap pages = (Bitmap)Image.FromFile(@"E:\\resultMFP.tif"))
            {
                var appDataPath = @"E:\Desktop\Wilde Testwiese\A Combini\asdsad";
                var filePath = Path.Combine(appDataPath, Path.GetRandomFileName() + ".tif");

                //save the first frame
                pages.Save(filePath, info, encoderParameters);

                //Save the second frame if any
                int frameCount1 = pages.GetFrameCount(FrameDimension.Page);
                if (frameCount1 > 1)
                {
                    for (int i = 1; i < frameCount1; i++)
                    {
                        encoderParameters.Param[0] = new EncoderParameter(enc, (long)EncoderValue.FrameDimensionPage);
                        pages.SelectActiveFrame(FrameDimension.Page, i);
                        pages.SaveAdd(pages, encoderParameters);
                    }
                }

            }```

**Comment by dlemstra** [comment]
The `MagickReadSettings` class has a `FrameIndex` property that can be used to set the start index and with `FrameCount` you can set the number of frames that you want to read.

**Reply by M4nju** [reply]
Yeah thats what i already figured out. But i want to create a multipage tiff file without having all pages inside the memory. Sorry if my post have been confusing.

So something like adding page by page to an existing file like in the Windows Bitmap code snippet.

**Reply by dlemstra** [reply]
It looks like you can append a tiff page to another tiff file. Have you tried that?

**Reply by M4nju** [reply]
Are you talking about adding the images to the MagickImageCollection? Yeah i know that this works but it comes with the explained problem that all pages of the tiff are inside the RAM:

```
            // The list of single page files which should be combined. Can be up to 400.
            List<string> singlePageFiles = new List<string>()
            {
                @"E:\Desktop\Wilde Testwiese\A Combini\TestRotationtest - Kopie (2).tif",
                @"E:\Desktop\Wilde Testwiese\A Combini\TestRotationtest - Kopie.tif",
                @"E:\Desktop\Wilde Testwiese\A Combini\TestRotationtest.tif",
                @"E:\Desktop\Wilde Testwiese\A Combini\2.tif"

            };

            using (MagickImageCollection multipageTif = new MagickImageCollection())
            {
                foreach (string singePageFile in singlePageFiles)
                {
                        // Every time a image is added, the ram usage will increase by the image size, as it is loaded into the ram
                       // It will only be Disposed after the multipageTiff is disposed. With large files the ram usage extends
                       // dramaticly.
                       multipageTif.Add(singePageFile);
                }

                // If there are many single page tiffs the ram is now at a extremly high usage, because all single page
                // files are loaded.
                multipageTif.Write(@"E:\Desktop\Wilde Testwiese\A Combini\Result.tif", MagickFormat.Tiff);
            }
```

**Reply by M4nju** [reply]
Well i will be using the ResourceLimitation for the RAM memory to counter the excessive usage. Not the perfect solution but i will work. Still thank you for the replies.

**Comment by M4nju** [comment]
Well i will be using the ResourceLimitation for the RAM memory to counter the excessive usage. Not the perfect solution but i will work. Still thank you for the replies.

**Reply by dlemstra** [reply]
I have had no time yet to respond to this but I wonder if you could not write two tiff pages to the same `FileStream`. Where you first write the first frame to the file and then open the file stream and append the second file. Have not had time to test if this works.

**Reply by M4nju** [reply]
Okay no problem .. didn't want to stress you out or something. I found this solution for me for now. If you have the time and patience i would love to hear from you. The solution sounds interesting and I will look into it as soon as I have time 👍

**Reply by M4nju** [reply]
Okay i tried it with an filestream but couldn´t get it to work properly. If i ad an image to the collection, save it to file stream and than dispose the image, the collection is also disposed. I tried creating a new collection and this increases the file size but does not append another page to the tiff:

``` 
using (MagickImageCollection collection = new MagickImageCollection())
            {
                using (MagickImage imageNew = new MagickImage(@"E:\Desktop\Wilde Testwiese\A Combini\1.tif"))
                using (MagickImage imageNew2 = new MagickImage(@"E:\Desktop\Wilde Testwiese\A Combini\3.tif"))
                {
                    collection.Add(imageNew);
                    collection.Add(imageNew2);
                    using (FileStream streami = new FileStream(@"E:\Desktop\Wilde Testwiese\A Combini\FileStreamResult.tif", FileMode.Create))
                    {
                        collection.Write(streami, MagickFormat.Tiff);
                    }
                }
            }

            using (MagickImageCollection collection = new MagickImageCollection())
            {
                using (MagickImage imageNew3 = new MagickImage(@"E:\Desktop\Wilde Testwiese\A Combini\CPCP.tif"))
                {
                    collection.Add(imageNew3);
                    using (FileStream streami = new FileStream(@"E:\Desktop\Wilde Testwiese\A Combini\FileStreamResult.tif", FileMode.Append))
                    {
                        collection.Write(streami);
                    }
                }
                 
            }
```

**Reply by M4nju** [reply]
Just as an update, i am now working with libtiff.net to combine multipage images. First i save the single page images with ImageMagick using the YCbCr colorspace and than i append each page to the first using libtiff. (If your using YCbCr you need to write the encoded strip with libtiff else it does not work).
